### PR TITLE
charts: add TLS support for TiDB-Lightning and TiKV-Importer

### LINKED ISSUE #1631: Integrate with Lightning for TLS support
## Feature Request

**Is your feature request related to a problem? Please describe:**

https://github.com/pingcap/tidb-lightning/issues/262 is created for tidb-lightning to support TLS.
**Describe the feature you'd like:**

This feature is to trace the integration work in tidb-operator when the above enhancement is done.

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve?

Add TLS support for TiDB-Lightning and TiKV-Importer charts

close #1631

### What is changed and how does it work?

1. mount TLS secret if specified
2. add TLS args in the config file if specified

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 
   1. prepare data need to be imported into the TiDB cluster
   2. create a cluster Secrect for TiKV-Importer
     an YAML example with cert-manager is:
     ```yaml
    apiVersion: cert-manager.io/v1alpha2
     kind: Certificate
     metadata:
       name: ${cluster_name}-importer-cluster-secret
       namespace: ${namespace}
     spec:
       secretName: ${cluster_name}-importer-cluster-secret
       duration: 8760h # 365d
       renewBefore: 360h # 15d
       organization:
       - PingCAP
       commonName: "TiDB"
       usages:
         - server auth
         - client auth
       dnsNames:
       - "${cluster_name}-importer"
       - "${cluster_name}-importer.${namespace}"
       - "${cluster_name}-importer.${namespace}.svc"
       ipAddresses:
       - 127.0.0.1
       - ::1
       issuerRef:
         name: ${cluster_name}-tidb-issuer
         kind: Issuer
         group: cert-manager.io
     ```
   3. create a cluster Secrect for TiDB-Lightning
     an YAML example with cert-manager is:
     ```yaml
     apiVersion: cert-manager.io/v1alpha2
     kind: Certificate
     metadata:
       name: ${cluster_name}-lightning-cluster-secret
       namespace: ${namespace}
     spec:
       secretName: ${cluster_name}-lightning-cluster-secret
       duration: 8760h # 365d
       renewBefore: 360h # 15d
       organization:
       - PingCAP
       commonName: "TiDB"
       usages:
         - server auth
         - client auth
       dnsNames:
       - "${cluster_name}-lightning"
       - "${cluster_name}-lightning.${namespace}"
       - "${cluster_name}-lightning.${namespace}.svc"
       ipAddresses:
       - 127.0.0.1
       - ::1
       issuerRef:
         name: ${cluster_name}-tidb-issuer
         kind: Issuer
         group: cert-manager.io
     ```
   4. verify `${cluster_name}-tidb-client-secret` Secrect exists or create a similar one for TiDB-Lightning
   5. start a TiDB cluster with both `tlsCluster` and `tlsClient` enabled
   6. `helm install` `tikv-importer` chart with `tlsCluster.enabled: true`
   7. `helm install` `tidb-lightning` chart with `tlsCluster.enabled: true` and `tlsClient.enabled: true` (and with/without `tlsClient.secretName` set)
   8. verify the job of `tidb-lightning` is `Completed`
   9. verify the data have been imported into the TiDB cluster

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [x] Need to cherry-pick to the release branch 
- [x] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Add TLS support for TiDB-Lightning and TiKV-Importer charts
```

### CODE REVIEW

**Comment by DanielZhangQD** [comment]
@csuzhangxc could you please help confirm the following question?

- Does TiDB Lightning need to use the certs among the components and the certs between the TiDB server with Local backend and importer backend?
    - If TiDB Lightning needs to enable mTLS for itself, we need to ask users to create secrets as we have done for TiKV in the [doc](https://docs.pingcap.com/zh/tidb-in-kubernetes/dev/enable-tls-between-components) and mount it to the Lightning Pod.
    - If TiDB Lightning needs to use the cluster client certs to connect to the other components,  we need to mount the cluster client secret created in the [doc](https://docs.pingcap.com/zh/tidb-in-kubernetes/dev/enable-tls-between-components) to the Lightning Pod.  
    -  If TiDB Lightning needs to use the certs to connect to the TiDB server, we need to mount the TiDB client secret created in the [doc](https://docs.pingcap.com/zh/tidb-in-kubernetes/dev/enable-tls-for-mysql-client) to the Lightning Pod. In addition, users can create different TiDB client certs for different client apps, so we also have to support specific client secret for Lightning as we do for the other client apps in the [doc](https://docs.pingcap.com/zh/tidb-in-kubernetes/dev/enable-tls-for-mysql-client#%E7%AC%AC%E4%BA%8C%E6%AD%A5%E9%83%A8%E7%BD%B2-tidb-%E9%9B%86%E7%BE%A4).
- The same comments for the importer.

Please note that the secret name for the components except for the ones for different TiDB client apps and the keys in the secrets are fixed, which is what we are doing in the current TLS docs, so no need to make them configurable.

You can refer to what we do for TiDB cluster [here](https://github.com/pingcap/tidb-operator/blob/master/charts/tidb-cluster/values.yaml#L72) but may have to add two fields for TiDB Lightning.

File: charts/tidb-lightning/values.yaml [hunk 1]
```diff
@@ -59,6 +59,19 @@ targetTidbCluster:
...
+# To create this Secret: kubectl create secret generic <targetTidbCluster.name>-lightning-cluster-secret --namespace=<namespace> --from-file=tls.crt=<path/to/tls.crt> --from-file=tls.key=<path/to/tls.key> --from-file=ca.crt=<path/to/ca.crt>
+# tlsCluster:
+#   enabled: true
+
+# Whether enable the TLS connection with the TiDB MySQL protocol port.
+# if enabled, a Secret named `secretName` (if specified) or `${targetTidbCluster.name}-tidb-client-secret` must exist.
+# To create this Secret: kubectl create secret generic ${targetTidbCluster.name}-tidb-client-secret --namespace=<namespace> --from-file=tls.crt=<path/to/tls.crt> --from-file=tls.key=<path/to/tls.key> --from-file=ca.crt=<path/to/ca.crt>
+# tlsClient:
+#   enabled: true
+#   secretName: ${targetTidbCluster.name}-tidb-client-secret
```
Comment (csuzhangxc) [inline charts/tidb-lightning/values.yaml:73]: Should we need to rename it to `tlsClientSecretName` and move it out of `tlsClient`?

**Comment by csuzhangxc** [comment]
@DanielZhangQD PTAL again

File: charts/tidb-lightning/values.yaml (hunk 1 above)
Comment (DanielZhangQD) [reply charts/tidb-lightning/values.yaml]: @csuzhangxc I think we should rename it to `tlsClientSecretName` to keep consistent with others and just leave it under `tlsClient`.

File: charts/tidb-lightning/templates/_helpers.tpl [hunk 2]
```diff
@@ -22,6 +22,21 @@ config-file: |-
     {{- if .Values.config }}
 {{ .Values.config | indent 2 }}
     {{- end -}}
+    {{- if and .Values.tlsCluster .Values.tlsCluster.enabled }}
+  [security]
+  ca-path="/etc/lib/lightning-tls/ca.crt"
```
Comment (DanielZhangQD) [inline charts/tidb-lightning/templates/_helpers.tpl:27]: For the other components, we mount the secrets to `/var/lib/...`, I think we can keep consistent with the others.

File: charts/tidb-lightning/templates/_helpers.tpl [hunk 3]
```diff
@@ -22,6 +22,21 @@ config-file: |-
...
+  cert-path="/etc/lib/lightning-tls/tls.crt"
+  key-path="/etc/lib/lightning-tls/tls.key"
+    {{- end }}
+    {{- if and .Values.tlsClient .Values.tlsClient.enabled }}
+  [tidb.security]
+  ca-path="/etc/lib/tidb-client-tls/ca.crt"
```
Comment (DanielZhangQD) [inline charts/tidb-lightning/templates/_helpers.tpl:33]: ditto.

File: charts/tidb-lightning/templates/_helpers.tpl [hunk 4]
```diff
@@ -22,6 +22,21 @@ config-file: |-
...
+  cert-path="/etc/lib/tidb-client-tls/tls.crt"
+  key-path="/etc/lib/tidb-client-tls/tls.key"
+    {{- else }}
+  [tidb]
```
Comment (DanielZhangQD) [inline charts/tidb-lightning/templates/_helpers.tpl:37]: What if users configure other parameters in the `[tidb]` section in the `config` field of the [values.yaml](https://github.com/pingcap/tidb-operator/blob/master/charts/tidb-lightning/values.yaml#L82)?
According to the description [here](https://github.com/pingcap/tidb-lightning/blob/master/tidb-lightning.toml#L204), it seems that the default empty value can meet our requirement.

File: charts/tidb-lightning/templates/job.yaml [hunk 5]
```diff
@@ -138,6 +148,20 @@ spec:
...
       {{ end -}}
+      {{- if and .Values.tlsCluster .Values.tlsCluster.enabled }}
+      - name: lightning-tls
+        secret:
+          secretName: {{ .Values.targetTidbCluster.name }}-lightning-cluster-secret
+      {{- end }}
+      {{- if and .Values.tlsClient .Values.tlsClient.enabled }}
+      - name: tidb-client-tls
+        secret:
+        {{- if and .Values.tlsClient .Values.tlsClient.secretName }}
```
Comment (DanielZhangQD) [inline charts/tidb-lightning/templates/job.yaml:159]: `.Values.tlsClient` is already checked in L156.

File: charts/tikv-importer/templates/_helpers.tpl [hunk 6]
```diff
@@ -14,6 +14,12 @@ config-file: |-
     {{- if .Values.config }}
 {{ .Values.config | indent 2 }}
     {{- end -}}
+    {{- if and .Values.tlsCluster .Values.tlsCluster.enabled }}
+  [security]
+  ca-path="/etc/lib/importer-tls/ca.crt"
```
Comment (DanielZhangQD) [inline charts/tikv-importer/templates/_helpers.tpl:19]: ditto.

File: charts/tidb-lightning/values.yaml (hunk 1 above)
Comment (csuzhangxc) [reply charts/tidb-lightning/values.yaml]: renamed in https://github.com/pingcap/tidb-operator/pull/3598/commits/bf5d600fe320f6b29377a0bee798abae3a87fc85

File: charts/tidb-lightning/templates/_helpers.tpl (hunk 4 above)
Comment (csuzhangxc) [reply charts/tidb-lightning/templates/_helpers.tpl:37]: if the user set `tlsCluster.enabled: true`, then the default empty value will become `cluster`.

File: charts/tidb-lightning/templates/_helpers.tpl (hunk 2 above)
Comment (csuzhangxc) [reply charts/tidb-lightning/templates/_helpers.tpl]: fixed in https://github.com/pingcap/tidb-operator/pull/3598/commits/704b30b0f4c26e52ff41e1cf023b9ea1bc611ef1.

File: charts/tidb-lightning/templates/_helpers.tpl (hunk 3 above)
Comment (csuzhangxc) [reply charts/tidb-lightning/templates/_helpers.tpl]: fixed in https://github.com/pingcap/tidb-operator/pull/3598/commits/704b30b0f4c26e52ff41e1cf023b9ea1bc611ef1.

File: charts/tikv-importer/templates/_helpers.tpl (hunk 6 above)
Comment (csuzhangxc) [reply charts/tikv-importer/templates/_helpers.tpl]: fixed in https://github.com/pingcap/tidb-operator/pull/3598/commits/704b30b0f4c26e52ff41e1cf023b9ea1bc611ef1.

File: charts/tidb-lightning/templates/job.yaml (hunk 5 above)
Comment (csuzhangxc) [reply charts/tidb-lightning/templates/job.yaml]: fixed in https://github.com/pingcap/tidb-operator/pull/3598/commits/146540160a947ce7d2cc35e305abbe4c562dab60

File: charts/tidb-lightning/templates/_helpers.tpl (hunk 4 above)
Comment (csuzhangxc) [reply charts/tidb-lightning/templates/_helpers.tpl:37]: https://github.com/pingcap/tidb-lightning/blob/56bc32daad19b9dff10104c55300292de959fde3/lightning/config/config.go#L552-L571

Comment (DanielZhangQD) [reply charts/tidb-lightning/templates/_helpers.tpl:37]: OK, got it, considering the other configuration parameters in the `[tidb]` section, I think we can request users to set `tls: false` manually in the doc for this case.

Comment (csuzhangxc) [reply charts/tidb-lightning/templates/_helpers.tpl]: removed `[tidb]` section in https://github.com/pingcap/tidb-operator/pull/3598/commits/68200caf75269b0ce2e84b69ce41d22a28e8caf7.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**